### PR TITLE
RATEST-200: E2E Location management workflow test

### PR DESCRIPTION
**Pull request by Ndacyayisenga-droid**
E2E Location management workflow test
Issue link https://issues.openmrs.org/browse/RATEST-200
Ticket id RATEST-200

### CODE REVIEW

**Review by mherman22 — CHANGES_REQUESTED** [review]
you need to cd into `qaframework-bdd-tests` in the refapp-2x-location-management.yml file. something like **cd qaframework-bdd-tests**

refer to https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/.github/workflows/refapp-2x-condition.yml

**Comment by Ndacyayisenga-droid** [comment]
cc @kdaud @sherrif10

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationAttributeTypesPage.java [hunk 1]
```diff
@@ -33,6 +33,8 @@
 	private static final By CONFIRM_RETIRE_LOCATION_BUTTON = By.cssSelector("#retireLocationAttributeTypeForm button.confirm.right");
 	private static final By CANCEL_RETIRE_LOCATION_BUTTON = By.cssSelector("#retireLocationAttributeTypeForm button.cancel");
 	private static final By CANCEL_DELETE_LOCATION_BUTTON = By.cssSelector("#purgeLocationAttributeTypeForm button.cancel");
+	private static final By RETURN_TO_CONFIGURE_META_DATA_PAGE = By.cssSelector("#breadcrumbs > li:nth-child(2) > a");
+
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationAttributeTypesPage.java:37]: Remove this line

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationAttributeTypesPage.java [hunk 2]
```diff
@@ -82,8 +84,9 @@ public void selectPreferredHandler(String preferredHandler) {
 		selectFrom(PREFERRED_HANDLER_FIELD, preferredHandler);
 	}
 
-	public void enterHandlerConfiguration(String handlerConfiguarion) {
-		selectFrom(HANDLER_CONFIGURATION_FILED, handlerConfiguarion);
+	public void enterHandlerConfiguration(String handlerConfiguration) {
+		findElement(HANDLER_CONFIGURATION_FILED).clear();
+        findElement(HANDLER_CONFIGURATION_FILED).sendKeys(handlerConfiguration);
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationAttributeTypesPage.java:89]: Fix the indentation here. Ensure all the changes' indentation is the same.

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationTagsPage.java [hunk 3]
```diff
@@ -84,4 +85,8 @@ public void cancelRetireLocation() {
 		waitForElement(CONFIRM_RETIRE_LOCATION_BUTTON);
 		clickOn(CONFIRM_RETIRE_LOCATION_BUTTON);
 	}
+
+	public void returnToConfigureMetaDataHomePage() {
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationTagsPage.java:89]:
```suggestion
	public void returnToConfigureMetaDataPage() {
```

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        initiateWithLogin();
+    }
+
+    @After(RunTest.HOOK.SELENIUM_LOCATION_MANAGEMENT)
+    public void destroy() {
+        quit();
+    }
+
+    @Given("a user clicks on the configure metadata link from the home page")
+    public void clicksConfigureMetaDataLink(){
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java:59]:
```suggestion
    public void clickOnConfigureMetaDataLink(){
```

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        configureMetadataPage = homePage.goToConfigureMetadata();
+    }
+
+    @When("the system loads configure metadata link dashboard")
+    public void loadsConfigureMetaData(){
+        assertTrue(textExists("Configure Metadata"));
+    }
+
+    @And("a user clicks on Manage Location Attribute Types")
+    public void clickManageLocationAttributesTypes(){
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java:69]:
```suggestion
    public void clickOnManageLocationAttributesTypes(){
```

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    private static final String MIN_VALUE = "1";
+    private static final String DESCRIPTION = "This is the description";
+    private static final String NAME = "Noah";
+    private static final String LOCATION_NAME = "Bugema";
+    private static final String ADDRESS_ONE = "Bugema one";
+    private static final String ADDRESS_TWO = "Bugema two";
+    private static final String CITY = "Kampala";
+    private static final String STATE = "Central";
+    private static final String COUNTRY = "Uganda";
+    private ConfigureMetadataPage configureMetadataPage;
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java:41]: We can have spaces between these attributes. 

    
```

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    @Then("the form is saved")
+    public void theFormIsSaved(){
+        assertTrue(textExists("Noah"));
+        manageLocationAttributeTypesPage.returnToConfigureMetaDataPage();
+    }
+
+    @When("a user clicks on manage location tag")
+    public void clicksManageLocationTag(){
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java:103]:
```suggestion
    public void clicksOnManageLocationTag(){
```

File: qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test.properties [hunk 8]
```diff
@@ -2,7 +2,7 @@ webapp.url=https://qa-refapp.openmrs.org/openmrs
 login.username=admin
 login.password=Admin123
 login.location=Pharmacy
-webdriver=chrome
+webdriver=firefox
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test.properties:5]: Why changing from chrome to firefox ?

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationTagsPage.java (hunk 3 above)
Comment (Ndacyayisenga-droid) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationTagsPage.java:89]: > 

@sherrif10 there is an already exisiting method with such a name

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java (hunk 6 above)
Comment (Ndacyayisenga-droid) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java:41]: @sherrif10 what do you mean by spaces

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java:41]: Add line btn 40 and 41

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationTagsPage.java (hunk 3 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationTagsPage.java:89]: Which method is that.

**Comment by Ndacyayisenga-droid** [comment]
Cc @kdaud

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        manageLocationAttributeTypesPage.enterMinimumOccurs(MIN_VALUE);
+        manageLocationAttributeTypesPage.enterMaximumOccurs(MAX_VALUE);
+        manageLocationAttributeTypesPage.selectDatatype("Provider Datatype");
+        manageLocationAttributeTypesPage.enterDatatypeConfiguration("I have entered the data configuration");
+        manageLocationAttributeTypesPage.selectPreferredHandler("Provider Field Gen Datatype Handler");
+        manageLocationAttributeTypesPage.enterHandlerConfiguration("I have entered the handler configuration");
+    }
+
+    @Then("the user saves the form")
+    public void userSavesForm(){
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java:92]: Please this should have space. i
```suggestion
    public void userSavesForm() {
```Ensure all your method definition classes have space like the suggested above.

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationTagsPage.java (hunk 3 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationTagsPage.java:89]: Do you mind addressing this comment here.

**Comment by sherrif10** [comment]
Please cross-check all the left comments and addresss them.

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationTagsPage.java (hunk 3 above)
Comment (Ndacyayisenga-droid) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationTagsPage.java:89]: > Which method is that.

@sherrif10  this https://github.com/openmrs/openmrs-contrib-qaframework/blob/25dedb1043e0bac74946b8cf35b937d002d59734/qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationAttributeTypesPage.java#L120

**Comment by Ndacyayisenga-droid** [comment]
> Please cross-check all the left comments and addresss them.

@sherrif10 done

**Comment by sherrif10** [comment]
@Ndacyayisenga-droid  This  agreatly improved work. Let us also add this functonalities for retiring location found in this class https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/qaframework-legacy-tests/src/test/java/org/openmrs/contrib/qaframework/legacy/RetireLocationTest.java

**Comment by Ndacyayisenga-droid** [comment]
> @Ndacyayisenga-droid This agreatly improved work. Let us also add this functonalities for retiring location found in this class https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/qaframework-legacy-tests/src/test/java/org/openmrs/contrib/qaframework/legacy/RetireLocationTest.java

@sherrif10 do I need to create pages for the retire and delete functionality?

**Comment by sherrif10** [comment]
Nope thanks. You need to add these tests where they fit. pages come where its necesarry.

**Comment by Ndacyayisenga-droid** [comment]
> Nope thanks. You need to add these tests where they fit. pages come where its necesarry.

Its okay, was so first to type

**Comment by Ndacyayisenga-droid** [comment]
cc @sherrif10  @kdaud

**Comment by kdaud** [comment]
@Ndacyayisenga-droid have you taken a look at the build failure? https://github.com/openmrs/openmrs-contrib-qaframework/runs/5435639961?check_suite_focus=true

**Comment by Ndacyayisenga-droid** [comment]
cc @sherrif10  @kdaud

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationAttributeTypesPage.java [hunk 10]
```diff
@@ -113,4 +115,8 @@ public void retireLocation() {
 	public void cancelRetireLocation() {
 		clickOn(CANCEL_RETIRE_LOCATION_BUTTON);
 	}
+
+	public void returnToConfigureMetaDataPage() {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationAttributeTypesPage.java:119]: isn't the return type supposed to be a page?

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationTagsPage.java [hunk 11]
```diff
@@ -79,9 +81,17 @@ public void retireLocation() {
...
+		clickOn(RESTORE_RETIRE_LOCATION_ICON);
+	}
+
 	public void cancelRetireLocation() {
 		clickOn(CANCEL_RETIRE_LOCATION_ICON);
 		waitForElement(CONFIRM_RETIRE_LOCATION_BUTTON);
 		clickOn(CONFIRM_RETIRE_LOCATION_BUTTON);
 	}
+
+	public void returnToConfigureMetaDataHomePage() {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageLocationTagsPage.java:94]: is the return type supposed to be a page?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+import org.openmrs.contrib.qaframework.RunTest;
+
+import org.openmrs.contrib.qaframework.page.ManageLocationAttributeTypesPage;
+import org.openmrs.contrib.qaframework.page.LocationPage;
+import org.openmrs.contrib.qaframework.page.ManageLocationsPage;
+import org.openmrs.contrib.qaframework.page.AddEditLocationPage;
+import org.openmrs.contrib.qaframework.page.ManageLocationTagsPage;
+import org.openmrs.contrib.qaframework.page.ConfigureMetadataPage;
+
+public class LocationManagementSteps extends Steps {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java:29]: did you just forget to add an empty line here?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/05-location/locations.feature [hunk 13]
```diff
@@ -1 +1,28 @@
...
+    And a user saves add New location tag form
+    Then the New location tag form is saved
+    And the the user clicks on the retire location button
+    And the user cancels retire location
+    Then a user deletes location tag
+    # User story: Manager locations
+    When a user clicks on manager locations
+    Then the user click on add new location
+    And a user fills add new location form
+    Then the add new location form is saved
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/05-location/locations.feature:28]: add an empty line to cater for git collaborators

**Comment by Ndacyayisenga-droid** [comment]
cc @sherrif10  @kdaud

**Comment by sherrif10** [comment]
@Ndacyayisenga-droid  Kindly resolve the remaining conflicts. Let us know if you need more help.

**Comment by lumuchris256** [comment]
@Ndacyayisenga-droid 
Quick tip;
while working with forks, tis A good practise to make update the local and remote repos before pushing 
endavour to checkout a branch, not working on master

**Review by lumuchris256 — COMMENTED** [review]
@Ndacyayisenga-droid 
Quick tip;
while working with forks, tis A good practise to make update the local and remote repos before pushing 
endavour to checkout a branch, not working on master

**Comment by sherrif10** [comment]
@Ndacyayisenga-droid  Could you resolve merge conflicts, How far with the ticket, anything we can look into .

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+        manageLocationsPage.enterPostalCode("5678");
+        manageLocationsPage.selectLocationTag();
+    }
+
+    @Then("the add new location form is saved")
+    public void saveNewLocationForm(){
+        manageLocationsPage.saveLocation();
+    }
+}
+
```
Comment (pasindur99) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/LocationManagementSteps.java:174]: You just need to keep an empty line at the end of the file.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/05-location/locations.feature [hunk 15]
```diff
@@ -1 +1,29 @@
...
+
```
Comment (pasindur99) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/05-location/locations.feature:29]: Keep an empty line at the end of the file here as well.

**Review by pasindur99 — CHANGES_REQUESTED** [review]
Keep End of Line.

**Comment by sherrif10** [comment]
@Ndacyayisenga-droid  Resolve merge conflicts.

**Comment by sherrif10** [comment]
Hello @Ndacyayisenga-droid  Are you still working on this.

**Comment by sherrif10** [comment]
Unless you are not working on this, Feel free to un assigne yourself.

**Comment by sherrif10** [comment]
Closing this because @mherman22  is working on it https://github.com/openmrs/openmrs-contrib-qaframework/pull/428

**Pull request closed without merging**